### DISCUSSION TITLE
Overriding HTML templates

### DISCUSSION THREAD

**Discussion by MyNameIsNeXTSTEP** [mkdocs/mkdocs#3398] [Q&A] [answered]
## Preconditions
Mkdocs has build in HMTL template blocks, particularly for table-of-contents (toc). I work with Mkdocs-material, so it uses inside base.html the toc template [like this](https://github.com/squidfunk/mkdocs-material/blob/master/material/base.html#L177), which is declared [here](https://github.com/squidfunk/mkdocs-material/blob/master/material/partials/toc.html#L21) for example.

Right after that line of toc-item.html i want to insert my custom socials block:

{% include "partials/socials.html" %}

## Problem
Mkdocs is made such a way, that it uses the class of [nav block](https://github.com/squidfunk/mkdocs-material/blob/master/material/partials/toc.html#L8) of secondary type(class), and also it is used for the nested secondary level navigation tab in another toc block (left sidebar).

These two different blocks with similar inner nav-secondary classes you can see here:
![image](https://github.com/mkdocs/mkdocs/assets/79173403/4652ff8d-1749-4838-b0f9-d39f7c3e3bcf)
![image](https://github.com/mkdocs/mkdocs/assets/79173403/9edcdd6b-4ffa-47dd-92d1-756a512976f2)

When i insert the custom block to toc.html right after [md-nav__list](https://github.com/squidfunk/mkdocs-material/blob/master/material/partials/toc.html#L19) it gets embedded in two places and ruins the layout.

## Solutions used
- I tried to extend with {% extends "toc.html" %} default template layout and reconfigure classes for passing my template, didn't work
- I tried to copy all the base.html and disable/reconfigure special toc with different class, didn't work - whole page layout crashes.

## Question
I want to insert my custom template after right side toc block, how can i determine this place specifically from templates or in another way ?

**Comment by MyNameIsNeXTSTEP** [comment]
Question is open yet

**Comment by kamilkrzyskow** [comment, marked as answer]
That's really a question for the Material for MkDocs discussion board, as the theme defines the HTML partials that are rendered into the final website.

But anyway, the `toc.html` partial is included in both the content of the page and in the navigation, because toc in the navigation is being shown when the page is in mobile / portrait mode.

https://github.com/squidfunk/mkdocs-material/blob/929fc39dbe67dd5a2044636c81028270679d9213/material/templates/base.html#L177
https://github.com/squidfunk/mkdocs-material/blob/929fc39dbe67dd5a2044636c81028270679d9213/material/templates/partials/nav-item.html#L124

So basically to focus on the `toc` in the content on the right side, you need to override the `base.html`, there are multiple approaches to that:
1. Create a `main.html` file and extend `base.html` file and extend the `site_nav` block, as explained per the documentation 
    https://squidfunk.github.io/mkdocs-material/customization/#__codelineno-8-3 
    https://github.com/squidfunk/mkdocs-material/blob/929fc39dbe67dd5a2044636c81028270679d9213/material/templates/base.html#L157-L182
    Issue is that you have to copy over the whole thing and only change a part of it, I guess this part:
    a.
    ```html
    <div class="md-sidebar__inner">
       {% include "partials/toc.html" %}
       {% include "partials/socials.html" %}
    </div>
    ```
    or
    b.
    ```html
    <div class="md-sidebar__inner">
       {% include "partials/custom-toc.html" %}
    </div>
    ```
    Where the first suggestion adds the socials after the `nav` containing the toc, and the second suggestion uses a custom `toc.html` partial to avoid duplicate socials.
    This approach as it requires some copy pasting, could also require some maintenance, depending on how the Material theme updates the sections you copied.
2. Dynamically override the `base.html` using a hook. https://github.com/squidfunk/mkdocs-material/discussions/5060
    I've changed the approach a bit from the one in the guide, and haven't had time to update it. Should still be useful.
    This approach would immediately notify you of incompatible changes after a Material theme update, but could require similar or even more maintenance in this case.
3. Accept that you will have the socials partial included 2 times.
4. Maintain a custom `base.html`. Overall the worst option, just wrote it down to make the other options look better.

**Reply by MyNameIsNeXTSTEP** [reply]
Thanx, seems like the working solution to me, haven't try it yet.
When test it, i will report the result here.
